source/renderer/app/components/widgets/forms/PinCode.js

```diff
@@ -31,42 +38,198 @@ export default class PinCode extends Component<Props> {
+    const { value } = this.props;
+    const valueIsEmpty =
+      !value[key] || value[key] === '' || inputNewValue === '';
+    const valueHasChanged =
+      inputNewValue &&
+      inputNewValue.length === 1 &&
+      value[key] !== inputNewValue;
+    return valueIsEmpty || valueHasChanged;
+  };
+
+  inputIsMarked = (input: Field) => {
+    const { inputElement } = input;
+    const { current } = inputElement;
+    return current.selectionStart !== current.selectionEnd;
+  };
+
+  isInputFieldDisabled = (
+    index: number,
+    value: Array<string>,
+    disabled: boolean
+  ) => {
+    return (
+      disabled ||
+      (index !== 0 && (!value || !value[index - 1])) ||
+      (index > 1 && index <= 4 && !value[index - 2]) ||
+      (index === 3 && value[index - 3] === '')
+    );
+  };
 
-  onChange = (inputValue: ?number, key: number) => {
+  onChange = (inputValue: ?number | ?string, key: number) => {
     const { value, onChange } = this.props;
+    const { isBackSpace } = this.state;
     const inputNewValue =
-      inputValue !== null && inputValue !== undefined
+      inputValue !== null && inputValue !== undefined && !isNaN(inputValue)
         ? inputValue.toString()
         : '';
-
-    if (
-      !Object.prototype.hasOwnProperty.call(value, key) ||
-      value[key] === '' ||
-      inputNewValue === ''
-    ) {
+    if (this.inputHasNewValue(inputNewValue, key)) {
       const newValue = value;
-      newValue[key] = inputNewValue;
+      if (!isNaN(inputValue)) {
+        if (
+          isBackSpace &&
+          newValue[key] !== '' &&
+          inputNewValue === '' &&
+          this.focusKey !== key
+        ) {
+          newValue[key] = value[key];
+          const focusKey =
+            this.inputsRef[key] &&
+            this.inputsRef[key].inputElement.current.selectionStart === 0 &&
+            key > 2 &&
+            key < this.inputsRef.length
+              ? key - 1
+              : key;
+          setTimeout(() => {
+            const inputFieldRef = this.inputsRef[focusKey];
+            if (inputFieldRef && inputFieldRef.inputElement) {
+              inputFieldRef.focus();
+              if (inputFieldRef.props.value) {
+                inputFieldRef.inputElement.current.selectionStart = 1;
+                inputFieldRef.inputElement.current.selectionEnd = 1;
+              }
+              this.setState({ focusKeyChanged: false });
+            }
+          }, 50);
+        } else {
+          newValue[key] = inputNewValue;
+          this.setState({ isBackSpace: false, focusKeyChanged: false });
+        }
+      }
       if (onChange) {
-        onChange(newValue);
+        onChange(newValue, inputNewValue, key);
       }
       this.focusKey = key;
-      this.add = inputValue !== null && inputValue !== undefined;
+      this.add =
+        inputValue !== null && inputValue !== undefined && !isNaN(inputValue);
     }
   };
 
   componentDidUpdate() {
```

Comment:
Same here … there is a lot of low-level stuff going on that is hard to understand at first sight